SproxydKeysScan/DuplicateKeysIngestion.js

```diff
@@ -0,0 +1,279 @@
+                    if (!objectKey.includes('\u0000')) {
+                        // skip non-versioned objectKeys
+                        return;
+                    }
+
+                    const bucket = record.db;
+
+                    let json = null;
+                    try {
+                        json = JSON.parse(entry.value);
+                    } catch (err) {
+                        log.error('json corrupted');
+                        return;
+                    }
+
+                    // skip empty objects
+                    if (!json.location || !Array.isArray(json.location)) {
+                        return;
+                    }
+                    const sproxydKeys = json.location.map(loc => loc.key);
+                    extractedKeys.push({ objectKey, sproxydKeys, bucket });
+                });
+            }
+        });
+        log.info('processBatch succeeded');
+        return cb(null, extractedKeys);
+    }
+
+    /**
+     * Hands off keys to the processor and updates this.begin to mark next offset to read from.
+     * @param {Array.<ExtractedKey>} extractedKeys - Array of ExtractedKey instances.
+     * @param {function(err:Error, res:Boolean)} cb - callback. Response res is true if success.
+     * @returns {undefined}
+     * @callback called with (err:Error, res: Boolean).
+     */
+    updateStatus(extractedKeys, cb) {
+        for (const entry of extractedKeys) {
+            try {
+                this.processor.insert(entry.objectKey, entry.sproxydKeys, entry.bucket);
+            } catch (err) {
+                log.error('insert key failed in updateStatus', { err, entry });
+                return cb(err, null);
+            }
+        }
+
+        // if we go over cseq, start at cseq + 1 while waiting for new raft journal entries
+        this.begin = Math.min(this.limit + this.begin, this.cseq + 1);
+        log.info('updateStatus succeeded');
+        return cb(null, true);
+    }
+
+    /**
+     * Reads, processes, and updates status of one batch of objects from the the Raft Journal.
+     * If an error occurs at any point in the process, or if there are no new objects to be processed
+     * a timeout of 5 seconds is set before polling again.
+     * @param {function(err:Error, res:number)} cb - callback. Response res is a timeout in milliseconds.
+     * @returns {undefined}
+     * @callback called with (err:Error, res:number)
+     */
+
+    runOnce(cb) {
+        return waterfall([
+            next => this.setBegin(
+                err => {
+                    if (err) {
+                        log.error('in setBegin', { error: err.message });
+                        next(err);
+                    } else {
+                        next(null);
+                    }
+                }),
+            next => this.getBatch(
+                (err, res) => {
+                    if (err) {
+                        log.error('in getBatch', { error: err.message });
```

Comment:
You can skip logging (or debug log only) of the errors of type "416" (requested range not satisfiable) since those errors would occur as soon as the log has been processed entirely without any new entry to process.